### ISSUE TITLE
Raw Queries

### ISSUE THREAD

**Issue by developerium** [Nozbe/WatermelonDB#103]
Hi 

Is there anyway we can run custom queries on database?

Is there anyway we can access database directly?

**Comment by bsr203** [comment]
I used to be able to connect to the sqlite database through https://github.com/facebook/stetho

you can see the instructions at https://facebook.github.io/react-native/docs/debugging#debugging-with-stetho-http-facebookgithubio-stetho-on-android

But, not sure what changed, I cannot see my WatermelonDB database though `stetho` anymore, but other sqlite dbs. I like to know how people normally does it as well.

**Comment by radex** [comment]
> Is there anyway we can run custom queries on database?

Not currently. What exactly are you trying to achieve?

**Comment by MrTinyByte** [comment]
I want to make a report on several tables, a sum report on multiple joined tables.

**Comment by radex** [comment]
https://github.com/Nozbe/WatermelonDB/blob/master/src/adapters/sqlite/index.js#L162

this is the code that executes the query. The query will pass through `encodeQuery` to generate SQL. You could maybe extend `Query` or make a new `RawQuery` object or something like that to be able to pass raw SQL. Contributions welcome.

If you want a quick & dirty hack (but I warn you that it's very dirty, might not work in the future or at all, and you don't have observability, and it might be unsafe):

```js
import { NativeModules } from 'react-native'
import { sanitizeQueryResult } from '@nozbe/watermelondb/adapters/common'

const tag = database.adapter._tag
const bridge = NativeModules.DatabaseBridge

const dirtyRecords = await bridge.query(tag, 'select from ...')
const rawRecords = sanitizeQueryResult(dirtyRecords, appSchema.tables['name_of_table'])

const records = database.collections.get('name_of_table')._cache.recordsFromQueryResult(rawRecords)
```

does this help?

**Comment by radex** [comment]
@MrTinyByte hey, still interested in contributing to this? Did you try out the hack I described above?

**Comment by servocoder** [comment]
I was looking a solution for SQL `order` and `limit` support, and found the explanations why it's not an easy thing considering observables, e.g. #15. While this is true for observables there are cases when you just need to fetch some DB entity out of the observables context. My particular case is that I need to find first entity sorted by timestamp with `fetch()` method. If we can't get it with `{Q}` then raw queries would become a solution.

Thanks to @zobeirhamid we have PR #199 and, if it's ok, I would vote for it to be merged.

@radex take a look this PR please.

**Comment by sursma** [comment]
I am working heavy data intensive application,  need often  orderby limit,  max etc. ops,   either orderby and limit support or raw query support would be great!

**Comment by radex** [comment]
@sursma Take a look at the open PR :) It's been stuck for a while, maybe you can take it over and finish up so we can merge it?

**Comment by sursma** [comment]
@radex  I wrote a wrapper on top of raw query syntax its working well for me :-)

const rawQuery = async (db, tableName, sqlQuery) => {
  try {
    const tag = db.adapter._tag
    const bridge = NativeModules.DatabaseBridge
    const dirtyRecords = await bridge.query(tag, sqlQuery)
    const rawRecords = sanitizeQueryResult(dirtyRecords, mySchema.tables[tableName])
    const records = db.collections.get(tableName)._cache.recordsFromQueryResult(rawRecords)
    return records
  } catch (err) {
    return []
  }
}

**Comment by servocoder** [comment]
> ```js
> import { NativeModules } from 'react-native'
> import { sanitizeQueryResult } from '@nozbe/watermelondb/adapters/common'
> 
> const tag = database.adapter._tag
> const bridge = NativeModules.DatabaseBridge
> 
> const dirtyRecords = await bridge.query(tag, 'select from ...')
> const rawRecords = sanitizeQueryResult(dirtyRecords, appSchema.tables['name_of_table'])
> 
> const records = database.collections.get('name_of_table')._cache.recordsFromQueryResult(rawRecords)
> ```
> 
> does this help?

Doesn't work anymore.
Can you force PR #199 please?

**Comment by radex** [comment]
@servocoder should work with minor changes. #199 has been sitting there for a while, nearly done. I do encourage you to take over the PR with the finishing touches discussed there and contribute if you'd like to have this available :)

**Comment by servocoder** [comment]
I'd already have it done if I had time this :)
Anyhow I'm looking forward this feature

**Comment by radex** [comment]
shipped in 0.15
